wsd/DocumentBroker.cpp

```diff
@@ -3597,17 +3597,26 @@ bool DocumentBroker::sendUnoSave(const std::shared_ptr<ClientSession>& session,
     // Invalidate the timestamp to force persisting.
     _saveManager.setLastModifiedTime(std::chrono::system_clock::time_point());
 
+    static const bool forceBackgroundEnv = !!getenv("COOL_FORCE_BGSAVE");
+    constexpr std::size_t MaxFailureCountForBackgroundSaving = 2; // Give only 1 extra chance.
+
+    // Note: It's odd to capture these here, but this function is used from ClientSession too.
+    const bool autosave = isAutosave || (_unitWsd && _unitWsd->isAutosave());
+    const bool backgroundConfigured = (autosave && _backgroundAutoSave) || _backgroundManualSave;
+    const bool canBackground = forceBackgroundEnv || (!finalWrite && backgroundConfigured);
+    const bool background = canBackground && _saveManager.lastSaveSuccessful() &&
+                            _saveManager.saveFailureCount() < MaxFailureCountForBackgroundSaving;
+
     std::ostringstream oss;
     // arguments init
     oss << '{';
 
-    if (dontTerminateEdit)
-    {
-        // We do not want save to terminate editing mode if we are in edit mode now.
-        //TODO: Perhaps we want to terminate if forced by the user,
-        // otherwise autosave doesn't terminate?
-        oss << "\"DontTerminateEdit\" : { \"type\":\"boolean\", \"value\":true }";
-    }
+    // We do not want save to terminate editing mode if we are in edit mode now.
+    // We want to terminate if forced by the user, otherwise autosave doesn't terminate.
+    dontTerminateEdit = dontTerminateEdit && background == false;
```

Comment:
Would be good to add that "we always force terminating edit when saving in the background."